cpp/src/dictionary/replace.cu

```diff
@@ -0,0 +1,242 @@
+                                        rmm::mr::device_memory_resource* mr,
+                                        cudaStream_t stream)
+{
+  auto const input_view = column_device_view::create(input, stream);
+  auto const d_input    = *input_view;
+  auto predicate        = [d_input] __device__(auto i) { return d_input.is_valid(i); };
+
+  auto input_pair_iterator = make_nullable_index_iterator<true>(input);
+  return cudf::detail::copy_if_else(true,
+                                    input_pair_iterator,
+                                    input_pair_iterator + input.size(),
+                                    replacement_iter,
+                                    predicate,
+                                    mr,
+                                    stream);
+}
+
+}  // namespace
+
+/**
+ * @copydoc cudf::dictionary::detail::replace_nulls(cudf::column_view const&,cudf::column_view
+ * const&,rmm::mr::device_memory_resource*,cudaStream_t)
+ */
+std::unique_ptr<column> replace_nulls(dictionary_column_view const& input,
+                                      dictionary_column_view const& replacement,
+                                      rmm::mr::device_memory_resource* mr,
+                                      cudaStream_t stream)
+{
+  if (input.size() == 0) { return cudf::empty_like(input.parent()); }
+  if (!input.has_nulls()) { return std::make_unique<cudf::column>(input.parent()); }
+  CUDF_EXPECTS(input.keys().type() == replacement.keys().type(), "keys must match");
+  CUDF_EXPECTS(replacement.size() == input.size(), "column sizes must match");
+
+  // first combine the keys so both input dictionaries have the same set
+  auto input_matched    = dictionary::detail::add_keys(input, replacement.keys(), mr, stream);
+  auto const input_view = dictionary_column_view(input_matched->view());
+  auto repl_matched     = dictionary::detail::set_keys(
+    replacement, input_view.keys(), rmm::mr::get_current_device_resource(), stream);
+  auto const repl_view = dictionary_column_view(repl_matched->view());
+
+  // now build the new indices by doing replace-null using the updated input indices
+  auto const input_indices = input_view.get_indices_annotated();
+  auto const repl_indices  = repl_view.get_indices_annotated();
+  auto new_indices =
+    repl_indices.has_nulls()
+      ? replace_indices(input_indices, make_nullable_index_iterator<true>(repl_indices), mr, stream)
+      : replace_indices(
+          input_indices, make_nullable_index_iterator<false>(repl_indices), mr, stream);
+
+  auto const indices_type = new_indices->type();
+  auto const indices_size = new_indices->size();
+  auto const null_count   = new_indices->null_count();
+  auto contents           = new_indices->release();
+  auto const new_type     = get_indices_type_for_size(input_view.keys().size());
+  // create the indices for the output column
+  auto indices_column = [&] {
+    if (new_type.id() == cudf::type_id::UINT32) {
```

Comment:
Can you add a comment explaining why this only happens for UINT32 columns?  I don't fully understand what is happening.